### PR TITLE
add optional schema migrations; default to olm.bundle.object instead of olm.csv.metadata

### PR DESCRIPTION
**Pull request by grokspawn**
**Description of the change:**
Introduces the concept of optional schema migrations to `opm` available to the render and migrate actions, and template expansions (basic and semver).  This includes a sequencing and extension mechanism since we presume that successive schema migrations might be dependent on the successful execution of prior ones. 

Schema migrations represent a level, are explicitly requested by name, and the sequence of all migrations up to and including the named one will be executed in sequence.

If any schema migration fails, it is immediately terminal to the sequence.

The existing `--migrate` flag for `opm render` is equivalent to specifying the last possible `--migrate-level`:
```sh
./bin/opm render --help
Generate a stream of file-based catalog objects to stdout from the provided
catalog images, file-based catalog directories, bundle images, and sqlite
database files.

...

      --migrate                Perform all available schema migrations on the rendered FBC

...
```

Information on available migrations can be obtained from the help for associated migrate/render/semver/basic commands, for e.g.:
```sh
./bin/opm render --help
Generate a stream of file-based catalog objects to stdout from the provided
catalog images, file-based catalog directories, bundle images, and sqlite
database files.

...
      --migrate-level string   Name of the last migration to run (default: none)

                               The migrator will run all migrations up to and including the selected level.

                               Available migrators:
                                 - none                          : do nothing
                                 - bundle-object-to-csv-metadata : migrates bundles' "olm.bundle.object" to "olm.csv.metadata"
...
```

**Motivation for the change:**
When we introduced the new [bundle metadata type](https://github.com/operator-framework/operator-registry/pull/1094) `olm.csv.metadata` and made it the default, we started hearing of authors having to select non-latest versions of `opm` in order to preserve their pipelines' functionality.  This is an anti-pattern, since we want to ensure that authors can safely use versions of `opm` that have the most fixes. 

**Reviewer Checklist**
- [ ] Implementation matches the proposed design, or proposal is updated to match implementation
- [ ] Sufficient unit test coverage 
- [ ] Sufficient end-to-end test coverage
- [ ] Docs updated or added to `/docs` 
- [ ] Commit messages sensible and descriptive

### CODE REVIEW

**Comment by joelanford** [comment]
> The existing --migrate flag for opm render is deprecated, and will exit with an error if used

That sounds like a breaking change. We _could_ make `--migrate` and `--migrate-level` mutually exclusive. If we did that, I think we would need to decide what `--migrate` means. I can think of two options:
1. It means the exact same thing as `--migrate-level=bundle-object-to-csv-metadata` (because that's what it means right now).
2. It means "do all of the available migrations". That's what I had imagined it would mean when we introduced it.

File: alpha/action/migrations/migrations.go
Comment (joelanford) [inline alpha/action/migrations/migrations.go:79]: Include `migration.Help()` here?

File: alpha/action/migrations/000_bundle_object_to_csv_metadata.go
Comment (joelanford) [inline alpha/action/migrations/000_bundle_object_to_csv_metadata.go:12]: This should probably be unexported now that I'm looking closer at it.

File: alpha/action/render.go
Comment (joelanford) [inline alpha/action/render.go:427]: Call `mobj.Migrate()` here (it encapsulates this loop).

File: alpha/property/property.go
Comment (joelanford) [inline alpha/property/property.go:262]: This looks identical to `MustBuildBundleObject` above. Do we need both?

File: alpha/template/semver/types.go
Comment (joelanford) [inline alpha/template/semver/types.go:15]: Nit: I see some fields called `MigrateLevel` and others called `MigrationLevel`. Should we make those consistent?

Only exception to this rule I could see if the `Migrate.Level` field, which makes sense to me to be named differently to avoid stuttering.

File: cmd/opm/alpha/render-graph/cmd.go
Comment (joelanford) [inline cmd/opm/alpha/render-graph/cmd.go:66]: Doesn't this mean "run no migrations"?

File: cmd/opm/migrate/cmd.go
Comment (joelanford) [inline cmd/opm/migrate/cmd.go:57]: What does this do? I understand at a high level that it helps with shell completion when a user is tabbing in their CLI after typing `--migrate <TAB>`.

But what happens with that? Will a user be able to tab through the names of each level?

File: alpha/property/property.go
Comment (grokspawn) [reply alpha/property/property.go:262]: Vestigial.  Cleaned up.

File: cmd/opm/alpha/render-graph/cmd.go
Comment (grokspawn) [reply cmd/opm/alpha/render-graph/cmd.go:66]: Good catch.  I didn't write that, and `opm render-graph` does not care about bundle metadata, so we shouldn't set anything.

File: cmd/opm/migrate/cmd.go
Comment (grokspawn) [reply cmd/opm/migrate/cmd.go:57]: This was a false start.  I'd hoped that it would allow us to provide a list of available migrators as a tab-completion, but it doesn't work except for bash, and I'm generally disenchanted with it.  
I left in, in case someone was going to point out my error, since I exhausted search results. 
I'll kill it, at least for now.

**Comment by grokspawn** [comment]
> > The existing --migrate flag for opm render is deprecated, and will exit with an error if used
> 
> That sounds like a breaking change. We _could_ make `--migrate` and `--migrate-level` mutually exclusive. If we did that, I think we would need to decide what `--migrate` means. I can think of two options:
> 
>     1. It means the exact same thing as `--migrate-level=bundle-object-to-csv-metadata` (because that's what it means right now).
> 
>     2. It means "do all of the available migrations". That's what I had imagined it would mean when we introduced it.

This is a breaking change as it is, and I tried to be really up-front to it.  I wanted something more elegant for the default-runs-all case than [this](https://github.com/operator-framework/operator-registry/pull/1384/commits/9e68d1c63e2303bcc99b16fadae9f87b33beb9bb#diff-e8b54ecee4543731d84f335c27814412b3419c50c057696d104012ede0b981f8R44).  There's also the argument to be made that if using the deprecation flag is an easy way to 'select all' then it would be the most common interaction users will have, and why would we want to deprecate it (versus providing overlapping options).

Separately, I'm uncomfortable with the reliance on bare strings instead of untyped data for the migration levels, and I was going to spend some more time with it, but in all honesty that can be follow-on work after this merges.

**Comment by grokspawn** [comment]
NB: I haven't updated the PR description based on the recent decision to reverse what the migrate flag means to render, since I think we're on the same page here but not 100% certain.  I'll update that with the ultimate result of the review.

File: alpha/action/migrations/migrations.go
Comment (grokspawn) [inline alpha/action/migrations/migrations.go:24]: This is because of [this](https://github.com/spf13/pflag/issues/200), where expansion of the template for the usage string imparts specific meaning to the first set of backticks it encounters.

File: alpha/action/migrate_test.go
Comment (joelanford) [inline alpha/action/migrate_test.go:23]: This is basically "no migrations", right?

I think we talked about this, but I feel like these tests could get very cumbersome once we have more than 2 migrations.

What if, instead, these tests ran all the migrations, and then we have a separate set of unit tests for the `migrations` package?

File: alpha/action/migrate.go
Comment (joelanford) [inline alpha/action/migrate.go:16]: Thinking about this a little more, what if we put `*migrations.Migrations` here (and on the other render-based structs) instead of the level? That way:
- tests of the render-based structs could use a simple test-only `Migrations` struct to verify that the migrations are executed (it doesn't really matter _what_ those migrations are)
- Other users of the library can use other migrations if they want.

File: alpha/action/migrations/migrations.go
Comment (joelanford) [reply alpha/action/migrations/migrations.go:24]: Instead of doing this "magic" replacement, should we just have the original `help` string use `"` directly?

File: alpha/action/migrations/000_bundle_object_to_csv_metadata.go
Comment (joelanford) [inline alpha/action/migrations/000_bundle_object_to_csv_metadata.go:15]: Nit: If we wanted the names and descriptions of all the migrations in one place, we could do something like the following instead.

Here: get rid of the `newMigration` wrapping and directly declare the migration function:
```go
func migrateBundleObjectsToCSVMetadata(cfg *declcfg.DeclarativeConfig) error { ... }
```

In `migrations.go` in line 74 where we setup `allMigrations`, add the `newMigration` wrapping:
```go
newMigration("bundle-object-to-csv-metadata", `migrates bundles' "olm.bundle.object" to "olm.csv.metadata"`, migrateBundleObjectsToCSVMetadata)
```

File: alpha/action/migrations/migrations.go
Comment (joelanford) [reply alpha/action/migrations/migrations.go:57]: Or even something like 
```go
newMigrations("none", "no migration", func(_ *declcfg.DeclarativeConfig) error { return nil}))
```

File: alpha/action/migrations/migrations.go
Comment (joelanford) [inline alpha/action/migrations/migrations.go:109]: We can use `migration.Token()` without other changes:
```suggestion
		fmt.Fprintf(tabber, "  - %s\t: %s\n", migration.Token(), migration.Help())
```

Go is able to figure out that `MigrationToken` is a string, so when it interprets it with `%s` formatting, it just uses the raw underlying string. If, in the future we need a special "token" to "name" function for some reason, I think we can define `String()` of the `MigrationToken` type and put whatever translation logic we want.

Comment (joelanford) [reply alpha/action/migrations/migrations.go:109]: If we do this, I think we can remove the `name` field and `Name()` method related to a migration.

File: alpha/action/migrations/migrations.go
Comment (joelanford) [inline alpha/action/migrations/migrations.go:69]: I feel like we should be able to incorporate this function into `NewMigrations` (or make it unexported and call it from there), especially if you agree that it makes sense for the `Render` struct (and others) to accept a *migrations.Migrations` directly, as I suggested elsewhere.

File: alpha/action/migrate_test.go
Comment (joelanford) [reply alpha/action/migrate_test.go:23]: (actually see my other comment about putting `*migrations.Migrations` on these structs and then just defining a dummy test `Migrations` instead)

File: alpha/action/migrations/000_bundle_object_to_csv_metadata.go
Comment (grokspawn) [reply alpha/action/migrations/000_bundle_object_to_csv_metadata.go:15]: LOL!  I /almost/ did that already, but I was making a concerted effort to not make 'nice to have' changes in my update, in order to get this in sooner.  
But now I think there's no excuse not to make the changes. 😁

File: alpha/action/migrate_test.go
Comment (grokspawn) [reply alpha/action/migrate_test.go]: There's a nuance here which is not the same as elsewhere.  I agree that we ought to be clear that what we're testing here is all/nothing and not incremental.  But in this case these are just coincidentally using identifiers that align with the current [0...1] options. 
We're using it in testdata construction to align with the selected all/none migrations. 
I've renamed/retyped this area to make it hopefully more clear.  WDYT?

**Comment by grokspawn** [comment]
/hold 
still working on the migrations_test in alpha/action/migrations.  Just pushed up to share a reference.

File: alpha/action/migrate_test.go
Comment (grokspawn) [reply alpha/action/migrate_test.go]: Adjusted this as we discussed to 
1. test only the none/arbitrary migrations cases, to verify that migrations-as-a-mechanism works;
2. introduce unit tests in actions.migrations which allow us to evaluate our sequence of migrators to protect against surprises along the processing chain

File: alpha/action/migrate_test.go
Comment (joelanford) [inline alpha/action/migrate_test.go:187]: I very well could be misunderstanding this code, but this will fail if `migrationPre` is ever non-zero, right?

And the only reason it doesn't fail now is because there's only one spec that increments the counter?

I'm having trouble conjuring an approach that remains simple and overcomes this problem, but I'll clone and tinker later.

File: alpha/action/migrations/migrations.go
Comment (joelanford) [inline alpha/action/migrations/migrations.go:94]: Now that we `validateName`, it seems like this is unreachable code? Should we drop this (or validateName)?

File: alpha/action/migrations/migrations_test.go
Comment (joelanford) [inline alpha/action/migrations/migrations_test.go:25]:
```suggestion
	evaluators := map[MigrationToken]func(*declcfg.DeclarativeConfig) error{
```

File: alpha/action/migrations/migrations_test.go
Comment (joelanford) [inline alpha/action/migrations/migrations_test.go:42]: WDYT about adding `input` and `expectedOutput` struct fields here instead of the `evaluators` indirection?

If we did that, the test code would turn into something like:
```
assert.NoError(t, migators.Migrate(t.input))
assert.Empty(t, cmp.Diff(t.input, t.expectedOutput))
```

File: alpha/action/migrations/migrations_test.go
Comment (joelanford) [inline alpha/action/migrations/migrations_test.go:33]: ?
```suggestion
			if diff := cmp.Diff(*d, csvMetadataCatalogFBC); diff != "" {
```

File: alpha/action/migrations/migrations.go
Comment (grokspawn) [reply alpha/action/migrations/migrations.go:94]: Yeah, this is dead, but my intention was to throw in validateName temporarily to validate the tokenizing, and evidently forgot to replace.  I don't think we gain much efficiency from skipping one loop, but we'll have it now. 😉

File: alpha/action/migrations/migrations_test.go
Comment (grokspawn) [reply alpha/action/migrations/migrations_test.go:60]: I was thinking to eval how golang performed a shallow copy by doing the value assignment, but... it seemed to work.  I definitely don't want to pollute the test sources, so I'll still func-ionalize it even though it appeared to be doing OK.

Comment (joelanford) [reply alpha/action/migrations/migrations_test.go:60]: yeah, you'll get a different `declcfg.DeclarativeConfig` object, but the copy and the original would have the same underlying slices.

File: alpha/action/migrate_test.go
Comment (grokspawn) [reply alpha/action/migrate_test.go:187]: Yes, this does an explicit compare against the counter value in the test spec... which will only work when the count of TCs attempting this feat is one.  
It should be verifying that the count went up by the amount in the test spec, instead:
```golang
require.Equal(t, migrationCounter, migrationPre+s.migrationCount)
```

File: alpha/action/migrations/migrations_test.go
Comment (grokspawn) [reply alpha/action/migrations/migrations_test.go:33]: This revealed that I was missing bundle.Image in my testdata, which was an oversight that makes the whole scenario collapse... so well spotted!

File: alpha/action/migrations/migrations_test.go
Comment (grokspawn) [reply alpha/action/migrations/migrations_test.go:42]: evaluators solves the problem of validating the sequence of migrators indicated by a migration label, whereas a test case may invoke multiple migrators (since the migrators label indicates a sequence [1..n] rather than a single value).  It's one-to-one with the available migrators so we can attest to no side effects as we progress through a migrator chain, not one-to-spec. 
i.e. 
 input --> none(unmutated) --> csvMetadata(mutated) --> futureMigration(mutatedagain).

File: pkg/registry/registry_to_model_test.go
Comment (joelanford) [inline pkg/registry/registry_to_model_test.go]: Delete this entirely?

**Review by joelanford — APPROVED** [review]

**Pull request merged**
